### PR TITLE
Option to hide agenda items

### LINKED ISSUE #468: Feat: make unlisted events
It would be nice to be able to make unlisted agenda events.
Currently people put them one year back etc.

### PR DESCRIPTION
**Pull request by jvbaast**
Adds the option to hide agenda items to the creation page. People can specify until when they want their agenda item to be hidden and can choose between: "Show immediately", "Until 1,2,3,4 weeks before" and "Indefinitely".

Has the potential to break the agenda page, so would be nice if a second person also tested this locally with some edge cases.

### CODE REVIEW

File: app/Http/Controllers/Api/AgendaController.php [hunk 1]
```diff
@@ -42,6 +44,24 @@ public function getAgenda(Request $request)
             $endDate = Carbon::createFromFormat('d-m-Y', $endDate)->endOfDay();
             $agendaItemQuery->where('startDate', '<=', $endDate);
         }
+        
+        # If user is not an administrator, filter out the hidden agenda items
+        if (Auth::user() == null || !Auth::user()->hasBackendRights()) {
```
Comment (ThijsEgberts) [inline app/Http/Controllers/Api/AgendaController.php:49]: I would make it so the unlisted agenda items are hidden for everyone, especially since a significant portion of active ESAC members have some form of backend rights

File: app/Http/Controllers/Api/AgendaController.php [hunk 2]
```diff
@@ -42,6 +44,24 @@ public function getAgenda(Request $request)
...
+            $agendaItemQuery->where('hidden', '<>', 404)->where('hidden', 0)
+                ->orWhere(function ($query) {
+                    $query->where('hidden', 1)->where('startDate', '<', Carbon::now()->addDays(8));
```
Comment (ThijsEgberts) [inline app/Http/Controllers/Api/AgendaController.php:52]: I feel like it would make more sense to do this as `'<=', Carbon::now()->addDays(7)`

File: database/factories/AgendaItemFactory.php [hunk 3]
```diff
@@ -22,5 +22,6 @@
         'subscription_endDate' => $faker->date,
         'image_url' => $faker->imageUrl,
         'climbing_activity' => $faker->boolean,
+        'hidden' => $faker->randomElement([0,1,2,3,4,404]),
```
Comment (ThijsEgberts) [inline database/factories/AgendaItemFactory.php:25]: Not super important, but it might be nice to make variants of the factory for the different unlist options because this factory is mainly used for tests and then I could imagine unlisting could cause unexpected behaviour.
See https://laravel.com/framework/docs/10.x/eloquent-factories#factory-states for the documentation. It's not super clear, but you can also see an example of what I mean in `UserFactory.php`

File: database/schema/mysql-schema.sql [hunk 4]
```diff
@@ -31,6 +31,7 @@ CREATE TABLE `agenda_items` (
   `created_at` timestamp NULL DEFAULT NULL,
   `updated_at` timestamp NULL DEFAULT NULL,
   `climbing_activity` tinyint(1) NOT NULL DEFAULT '1',
+  `hidden` int unsigned NOT NULL DEFAULT 0,
```
Comment (ThijsEgberts) [inline database/schema/mysql-schema.sql:34]: Please note that in Laravel we usually use migrations for mutating the db schema. This `schema.sql` file is essentially a bunch of migrations that are compressed for performance reasons.
See this documentation page: https://laravel.com/framework/docs/10.x/migrations

File: resources/lang/en/hide_agenda.php [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+return [
+    0 => "Show immediately",
+    1 => "Until 1 week before",
+    2 => "Until 2 weeks before",
+    3 => "Until 3 weeks before",
+    4 => "Until 4 weeks before",
+    404 => "Indefinitely"
```
Comment (ThijsEgberts) [inline resources/lang/en/hide_agenda.php:9]: I would maybe rename indefinitely to "hidden" just for extra clarity.
I would also change `404` to `5` to keep the format consistent with php arrays. (In case someone iterates over it and expects 404 to be 5)

**Comment by ThijsEgberts** [comment]
I have a design suggestion: maybe it would be nice to visually mark the hidden agenda items in the management `index` view, ie by making the agenda item title gray and prepending a hidden eye icon or something like that

**Review by ThijsEgberts — CHANGES_REQUESTED** [review]
In general it looks nice, I do have some comments which you can see below.

File: app/Http/Controllers/Api/AgendaController.php (hunk 2 above)
Comment (jvbaast) [reply app/Http/Controllers/Api/AgendaController.php:52]: I tried that but it doesn't work for some reason, so this was my work-around

Comment (jvbaast) [reply app/Http/Controllers/Api/AgendaController.php]: Fixed it, there was an issue since we use UTC has the timezone in the config file, so I had to specify the timezone. Is this not an issue for other places where we generate dates?

File: app/Http/Controllers/Api/AgendaController.php [hunk 6]
```diff
@@ -42,6 +44,39 @@ public function getAgenda(Request $request)
...
+                    $query->where('hidden', 2)->where('startDate', '<=', Carbon::now('Europe/Amsterdam')->addDays(14));
+                })
+                ->orWhere(function ($query) {
+                    $query->where('hidden', 3)->where('startDate', '<=', Carbon::now('Europe/Amsterdam')->addDays(21));
+                })
+                ->orWhere(function ($query) {
+                    $query->where('hidden', 4)->where('startDate', '<=', Carbon::now('Europe/Amsterdam')->addDays(28));
+                });
+        }
+        else if (!Auth::user()->hasRole(Config::get('constants.Administrator'))) {
```
Comment (ThijsEgberts) [inline app/Http/Controllers/Api/AgendaController.php:64]: I wouldn't make it so that hidden agenda items made by other people are hidden for activity administrators because that would probably be annoying when working together on an agenda item as a committee

File: app/Http/Controllers/Api/AgendaController.php [hunk 7]
```diff
@@ -42,6 +44,39 @@ public function getAgenda(Request $request)
             $endDate = Carbon::createFromFormat('d-m-Y', $endDate)->endOfDay();
             $agendaItemQuery->where('startDate', '<=', $endDate);
         }
+        
+        # If user is not an administrator, or not on the beheer page, filter out the hidden agenda items
+        if (Auth::guest() || $request->input('beheer') != true) {
+            $agendaItemQuery->where('hidden', 0)
+                ->orWhere(function ($query) {
+                    $query->where('hidden', 1)->where('startDate', '<=', Carbon::now('Europe/Amsterdam')->addDays(7));
```
Comment (ThijsEgberts) [inline app/Http/Controllers/Api/AgendaController.php:52]:
```suggestion
$query->where('hidden', 1)->where('startDate', '<=', Carbon::now('Europe/Amsterdam')->addWeeks(1));
```
I forgot in my previous review, but `Carbon` actually has an `addWeek()` function. I don't know if you tried it before, but this might be nice to use

File: app/Http/Controllers/Api/AgendaController.php [hunk 8]
```diff
@@ -42,6 +44,39 @@ public function getAgenda(Request $request)
             $endDate = Carbon::createFromFormat('d-m-Y', $endDate)->endOfDay();
             $agendaItemQuery->where('startDate', '<=', $endDate);
         }
+        
+        # If user is not an administrator, or not on the beheer page, filter out the hidden agenda items
+        if (Auth::guest() || $request->input('beheer') != true) {
```
Comment (ThijsEgberts) [inline app/Http/Controllers/Api/AgendaController.php:49]: If you want to keep this condition, I would change it to `!administrator || !beheer` to prevent some code duplication. If you want to keep the `->orWhere('createdBy', Auth::user()->id)` you could also add it to the end outside the if-else (I'm pretty sure).

File: app/Http/Controllers/Api/AgendaController.php [hunk 9]
```diff
@@ -42,6 +44,39 @@ public function getAgenda(Request $request)
...
+            $agendaItemQuery->where('hidden', 0)
```
Comment (ThijsEgberts) [inline app/Http/Controllers/Api/AgendaController.php:50]: Also another point I forgot about in my previous review: please make sure the order of operations is correct: in SQL `AND` takes precedence over `OR`, so 
```sql
someCondition AND "hidden" == 0 OR ("hidden" == 1 and "startdate" <= now+week) OR ("hidden" == 2 and "startdate" <= now+2week)
```
is executed as 
```sql
(someCondition AND "hidden" == 0) OR ("hidden" == 1 and "startdate" <= now+week) OR ("hidden" == 2 and "startdate" <= now+2week)
```
Which is probably not what we want, so you could use a closure to add parenthesis
```php
$agendaItemQuery->where(function (Builder $query) {
    $query->where('hidden', 1)
    ->orWhere(...)
    ->orWhere(...)
}
```
If you change this, don't forget to change it for `AgendaItemRepository.firstXItems()` as well

File: app/Repositories/AgendaItemRepository.php [hunk 10]
```diff
@@ -104,10 +106,23 @@ public function copy(AgendaItem $agendaItem): AgendaItem
     }
 
     public function getFirstXItems($limit)
-    {
+    {   
         return AgendaItem::query()
             ->with('getApplicationFormResponses')
-            ->whereDate('startDate', '>=', Carbon::now())
+            ->whereDate('startDate', '>=', Carbon::now('Europe/Amsterdam'))
```
Comment (ThijsEgberts) [inline app/Repositories/AgendaItemRepository.php:112]: See my comment on operator precedence in the controller

File: database/factories/AgendaItemFactory.php [hunk 11]
```diff
@@ -5,22 +5,84 @@
...
+            'text'=> fake()->text(),
+            'shortDescription' => fake()->word(),
+            'createdBy' => $user->id,
+            'category' => $category->id,
+            'startDate' => fake()->date(),
+            'endDate' => fake()->date(),
+            'subscription_endDate' => fake()->date(),
+            'image_url' => fake()->imageUrl(),
+            'climbing_activity' => fake()->boolean(),
+            'hidden' => fake()->randomElement([0,1,2,3,4,5]),
```
Comment (ThijsEgberts) [inline database/factories/AgendaItemFactory.php:28]: I would make the default agenda item non-hidden to prevent annoying bugs with tests

File: database/seeders/AgendaItemSeeder.php [hunk 12]
```diff
@@ -36,9 +36,10 @@ public function run()
         $agendaItem->text = 'Climbing at Neoliet North text';
         $agendaItem->shortDescription = 'Climbing at Neoliet North';
         $agendaItem->image_url = "";
-        $agendaItem->createdBy = 1;
+        $agendaItem->createdBy = 2;
         $agendaItem->category = 1;
         $agendaItem->application_form_id = 2;
+        $agendaItem->hidden = 5;
```
Comment (ThijsEgberts) [inline database/seeders/AgendaItemSeeder.php:42]: you could consider changing the title of this agenda item to clarify it should be hidden

File: resources/views/beheer/agendaItem/index.blade.php [hunk 13]
```diff
@@ -94,10 +95,19 @@
                 "dataSrc": "agendaItems"
             },
             columnDefs: [
-                {type: 'de_datetime', targets: 1},
                 {type: 'de_datetime', targets: 2},
```
Comment (ThijsEgberts) [inline resources/views/beheer/agendaItem/index.blade.php:98]: I can't place the comment on the right line, but can you change 
```php
"order": [[1, "asc"]]
```
to
```php
"order": [[2, "asc"]]
```
so the default sorting column is updated as well

File: app/Http/Controllers/Api/AgendaController.php (hunk 8 above)
Comment (ThijsEgberts) [inline app/Http/Controllers/Api/AgendaController.php:49]: Another note here that is not super important for this application, but using the query strings like this allows a logged in user without privileges to make a request to `/api/agenda?beheer=true` and get the hidden agenda items as well. Again, not important in this case, but since new people usually look at what is already there for how to do this it might be good practice to achieve this goal in another way. 
For example by making a separate endpoint for the beheer index since most of the options in this endpoint aren't relevant anyway.

**Review by ThijsEgberts — CHANGES_REQUESTED** [review]
Good work! Some more changes are needed, but it's definitely going in the right direction!

File: app/Http/Controllers/Api/AgendaController.php (hunk 2 above)
Comment (ThijsEgberts) [reply app/Http/Controllers/Api/AgendaController.php]: If I remember correctly, we usually do something like `->addHour()` when we display dates. I agree it's annoying, but if I remember correctly, somewhere in the docs they recommend to keep it at UTC to prevent other more technical issues or something

File: app/Http/Controllers/Api/AgendaController.php (hunk 8 above)
Comment (jvbaast) [reply app/Http/Controllers/Api/AgendaController.php]: I now changed it to check for activity administrator privileges in the same if-condition, so this shouldn't be an issue

File: app/Http/Controllers/Api/AgendaController.php (hunk 9 above)
Comment (jvbaast) [reply app/Http/Controllers/Api/AgendaController.php]: I think I fixed this, but not sure
